### PR TITLE
fix: Confirmation bridge verify backend

### PR DESCRIPTION
**Pull request by ejwessel**
## **Description**

Adds a middleware that verifies certain transactions from the Portfolio.

[Open in GitHub Codespaces](https://codespaces.new/MetaMask/metamask-extension/pull/25047?quickstart=1)

## **Related issues**

Fixes:

## **Manual testing steps**

1. Go to this page...
2.
3.

## **Pre-merge author checklist**

- [ ] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

**Comment by ejwessel** [comment]
I have read the CLA Document and I hereby sign the CLA

**Review by micaelae — DISMISSED** [review]

**Comment by ejwessel** [comment]
@metamaskbot update-policies

**Comment by rekmarks** [comment]
@metamaskbot update-policies

**Review by rekmarks — DISMISSED** [review]
LGTM.

**Review by micaelae — DISMISSED** [review]

**Comment by georgewrmarshall** [comment]
Hey @ejwessel, I noticed this has one approval so it seems like it could be ready for review. Could you please fill out the [PR description](https://github.com/MetaMask/contributor-docs/blob/main/docs/pull-requests.md)?

**Review by georgewrmarshall — DISMISSED** [review]
Approving on behalf of @MetaMask/design-system-engineers, but I'm not sure why we are tagged as codeowners since this PR does not touch the component-library folder. Please also seek review from the @MetaMask/extension-platform or another team with more context.

**Comment by ejwessel** [comment]
@rekmarks only thing left to decide on this is what address we want to use for signing.

File: app/scripts/metamask-controller.js
Comment (matthewwalsh0) [inline app/scripts/metamask-controller.js:5088]: Rather than coupling this to a specific origin, should we make it a little more generic and extensible via an array constant such as `TRANSACTION_VERIFICATION_ORIGINS`?

File: shared/constants/bridge.ts
Comment (matthewwalsh0) [inline shared/constants/bridge.ts:15]: Following the comment about a potential origin constant, should we also extend this into a map such as `TRANSACTION_VERIFICATION_SIGNER_BY_ORIGIN`?

In that case, we'd likely want to move these constants to a more generic file.

File: shared/constants/bridge.ts
Comment (matthewwalsh0) [inline shared/constants/bridge.ts:18]: `TRANSACTION_VERIFICATION_SIGNATURE_LENGTH` for clarity?

File: app/scripts/lib/tx-verification/tx-verification-middleware.ts
Comment (matthewwalsh0) [inline app/scripts/lib/tx-verification/tx-verification-middleware.ts:18]: The types and comments and constants are coupled to bridge, but if the mechanism could support any origin, should we instead frame it as a generic transaction verification that we currently just use for portfolio?

File: app/scripts/lib/tx-verification/tx-verification-middleware.ts
Comment (matthewwalsh0) [inline app/scripts/lib/tx-verification/tx-verification-middleware.ts:43]: We check this method in a few places such as the PPOM support, I wonder if it warrants a generic `ProviderMethod` enum or similar constant.

File: app/scripts/lib/tx-verification/tx-verification-middleware.ts
Comment (matthewwalsh0) [inline app/scripts/lib/tx-verification/tx-verification-middleware.ts:55]: Minor, but could we use the `Hex` type from `@metamask/utils`?

File: app/scripts/lib/tx-verification/tx-verification-middleware.ts
Comment (matthewwalsh0) [inline app/scripts/lib/tx-verification/tx-verification-middleware.ts:45]: We can skip this middleware if the method isn't a send transaction, but should we instead `end` if `isValidParams` fails since that would mean a bad extension could theoretically mutate a transaction request enough to fail these validations and skip the verification, but potentially still be processed by subsequent middleware and controllers?

File: app/scripts/lib/tx-verification/tx-verification-middleware.ts
Comment (matthewwalsh0) [inline app/scripts/lib/tx-verification/tx-verification-middleware.ts:56]: Should we avoid coupling the middleware to the entire controller and instead just pass a callback such `getGlobalChainId`?

We reference that property chain alot in `metamask-controller` so could remove lots of duplication also.

File: app/scripts/lib/tx-verification/tx-verification-middleware.ts
Comment (matthewwalsh0) [inline app/scripts/lib/tx-verification/tx-verification-middleware.ts:61]: Minor, but rather than reading the full property chain twice, could we instead use a variable such as:

```
const bridgeContractAddress = FIRST_PARTY_CONTRACT_NAMES['MetaMask Bridge'][chainId]?.toLowerCase();

if(!bridgeContractAddress || params.to.toLowerCase() !== bridgeContractAddress)
```

File: app/scripts/lib/tx-verification/tx-verification-middleware.ts
Comment (matthewwalsh0) [inline app/scripts/lib/tx-verification/tx-verification-middleware.ts:67]: Assuming the intent of this middleware is to protect portfolio to bridge transactions, then won't this return allow bad extensions to redirect bridge transactions to an alternate contract?

Could we verify all portfolio transactions?

File: app/scripts/lib/tx-verification/tx-verification-middleware.ts
Comment (matthewwalsh0) [inline app/scripts/lib/tx-verification/tx-verification-middleware.ts:78]: Minor, but could the above logic go into a `hashParams` function to break up this function?

File: app/scripts/lib/tx-verification/tx-verification-middleware.ts
Comment (matthewwalsh0) [inline app/scripts/lib/tx-verification/tx-verification-middleware.ts:82]: Would it be more explicit to call this `actualSigner` to match the signer constant?

File: app/scripts/lib/tx-verification/tx-verification-middleware.ts
Comment (matthewwalsh0) [inline app/scripts/lib/tx-verification/tx-verification-middleware.ts:86]: To clarify, if we throw here, will that mean the dApp gets the error but it never reaches the `TransactionController` and therefore no confirmation is even created?

File: app/scripts/metamask-controller.js
Comment (ejwessel) [reply app/scripts/metamask-controller.js:5088]: We could, but Portfolio is the only affected product, and there are no other supported products that require this guard (yet) that wouldn't already be a part of Portfolio. If there are services outside of portfolio that would warrant this change I'm for it.

File: shared/constants/bridge.ts
Comment (ejwessel) [reply shared/constants/bridge.ts:15]: If we had a truly generic solution (one that Uniswap is supposedly working on a proposal for). I think a mapping like that would fit really cleanly into some decentralized registry like ENS.
I have this as one of the potential solutions identified in this [solutions doc](https://docs.google.com/document/d/1yEH6VKyZf6sauMHTUs9WWZ7cv_d2QfE6q44B0qNpKdQ/edit), however the goal of this PR is to stop the [bleed](https://dune.com/queries/3733704/6279934) and iterate on a solution that further generalizes this concept.

File: app/scripts/lib/tx-verification/tx-verification-middleware.ts
Comment (ejwessel) [reply app/scripts/lib/tx-verification/tx-verification-middleware.ts:18]: Great point let me rethink how to structure this.

File: app/scripts/lib/tx-verification/tx-verification-middleware.ts
Comment (ejwessel) [reply app/scripts/lib/tx-verification/tx-verification-middleware.ts:55]: I will try.

File: app/scripts/lib/tx-verification/tx-verification-middleware.ts
Comment (rekmarks) [reply app/scripts/lib/tx-verification/tx-verification-middleware.ts:45]: I wrote most of this validation. I think we should fall back to our regular transaction handling if we can't validate the transaction in this middleware.

File: app/scripts/lib/tx-verification/tx-verification-middleware.ts
Comment (rekmarks) [reply app/scripts/lib/tx-verification/tx-verification-middleware.ts:56]: I think that would best be done in a follow-up: #25223

File: app/scripts/lib/tx-verification/tx-verification-middleware.ts
Comment (rekmarks) [reply app/scripts/lib/tx-verification/tx-verification-middleware.ts:86]: Not displaying a confirmation at all is the intended behavior. The error can be handled in the Portfolio.

File: app/scripts/lib/tx-verification/tx-verification-middleware.ts
Comment (ejwessel) [reply app/scripts/lib/tx-verification/tx-verification-middleware.ts]: We could, but it's technically incorrect. Just because an address comes out of verifyMessage() doesn't necessarily mean the address signed it. If the address is unexpected, it could indicate an invalid signature or altered parameters. In both cases, "actualSigner" would imply they signed, which isn't accurate.

File: shared/constants/bridge.ts
Comment (ejwessel) [reply shared/constants/bridge.ts:18]: If feel like that's pretty long, but if you feel strongly about it I'll change it.

File: app/scripts/metamask-controller.js
Comment (ejwessel) [reply app/scripts/metamask-controller.js:5088]: I have modified the code so that it still targets `origin === BaseUrl.Portfolio`, but that any experience could easily add an "experience to verify" with "trusted signers"

File: shared/constants/bridge.ts
Comment (ejwessel) [reply shared/constants/bridge.ts]: Similar to what I have written https://github.com/MetaMask/metamask-extension/pull/25047#discussion_r1635563623
There is now a map of trusted bridge signers broken down by experiences type

File: shared/constants/bridge.ts
Comment (ejwessel) [reply shared/constants/bridge.ts]: updated to something similar.

File: app/scripts/lib/tx-verification/tx-verification-middleware.ts
Comment (ejwessel) [reply app/scripts/lib/tx-verification/tx-verification-middleware.ts]: see restructure comment https://github.com/MetaMask/metamask-extension/pull/25047#discussion_r1635563623

File: app/scripts/lib/tx-verification/tx-verification-middleware.ts
Comment (ejwessel) [reply app/scripts/lib/tx-verification/tx-verification-middleware.ts:53]: I saw this as well, but decided to hold off on doing that within this PR.
I'd like to place such an enum within `eth-methods.ts`, however there would seem to be some overlap between `@metamask/keyring-api` and the `app.ts:MESSAGE_TYPE` and this new enum. In any event I opted for the `app.ts:MESSAGE_TYPE` and added two new entries.

File: app/scripts/lib/tx-verification/tx-verification-middleware.ts
Comment (matthewwalsh0) [inline app/scripts/lib/tx-verification/tx-verification-middleware.ts:75]: Some of the first party contracts use the same address on multiple chains, this means we can't assume an address correlates to a single chain ID.

Could we instead use a function such as `getExperience(address: Hex, chainId: Hex)` and verify we have a match?

We could also extend the `addrToExpMap` constant to include an array of chain IDs in the values, but that feels more complex and less readable.

File: app/scripts/lib/tx-verification/tx-verification-middleware.ts
Comment (matthewwalsh0) [inline app/scripts/lib/tx-verification/tx-verification-middleware.ts:92]: Ultra minor, but since this function **will** hash the param, should this be `hashParams`?

File: app/scripts/lib/tx-verification/tx-verification-middleware.ts
Comment (matthewwalsh0) [inline app/scripts/lib/tx-verification/tx-verification-middleware.ts:54]: ❤️

File: app/scripts/lib/tx-verification/tx-verification-middleware.ts
Comment (matthewwalsh0) [inline app/scripts/lib/tx-verification/tx-verification-middleware.ts:85]: Don't mean to complicate things, but is it likely we'll have different signers per chain for the same experience?

File: shared/constants/verification.ts
Comment (matthewwalsh0) [inline shared/constants/verification.ts:15]: Minor, but even though this is calculated, is this still a constant so uppercase and underscores?

File: app/scripts/lib/tx-verification/tx-verification-middleware.ts
Comment (ejwessel) [reply app/scripts/lib/tx-verification/tx-verification-middleware.ts:75]: > Some of the first party contracts use the same address on multiple chains, this means we can't assume an address correlates to a single chain ID.

you're right. That's an oversight on my part. I think a getExperience will resolve this without reverse look up.

File: app/scripts/lib/tx-verification/tx-verification-middleware.ts
Comment (ejwessel) [reply app/scripts/lib/tx-verification/tx-verification-middleware.ts]: It's possible, but it's very unlikely. Mainly because there's no point for a single service to handle individual keys across networks and instead it's likely that they'll have a single signer that signs across all networks.

In the event if we do I thought turning `TRUSTED_SIGNERS: Partial<Record<EXPERIENCES_TYPE, Hex>>` into 
`TRUSTED_SIGNERS: Partial<Record<EXPERIENCES_TYPE, Hex[]>>` would be relatively trivial.

File: shared/constants/verification.ts
Comment (ejwessel) [reply shared/constants/verification.ts:15]: I'm not sure I'm following... Do you mean `FIRST_PARTY_CONTRACT_NAMES` or the result of this?

File: app/scripts/lib/tx-verification/tx-verification-middleware.ts
Comment (matthewwalsh0) [reply app/scripts/lib/tx-verification/tx-verification-middleware.ts]: That makes sense, but given that scope, should we rename the PR since it's no longer the intent to update the confirmation?

File: shared/constants/verification.ts
Comment (matthewwalsh0) [reply shared/constants/verification.ts]: Apologies, I just meant should `addrToExpMap` be something like `EXPERIENCE_DATA_BY_ADDRESS`?

Comment (ejwessel) [reply shared/constants/verification.ts]: yeah i think the introduction of getExperience() doesn't make this relevant anymore

**Review by matthewwalsh0 — DISMISSED** [review]

**Review by rekmarks — APPROVED** [review]

**Review by darkwing — APPROVED** [review]

**Pull request merged**
